src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/StringLengthAttributeAdapter.cs

```diff
@@ -11,24 +11,32 @@ namespace Microsoft.AspNetCore.Mvc.DataAnnotations.Internal
 {
     public class StringLengthAttributeAdapter : AttributeAdapterBase<StringLengthAttribute>
     {
+        private readonly string _max;
+        private readonly string _min;
+
         public StringLengthAttributeAdapter(StringLengthAttribute attribute, IStringLocalizer stringLocalizer)
             : base(attribute, stringLocalizer)
         {
+            _max = Attribute.MaximumLength.ToString(CultureInfo.InvariantCulture);
+            _min = Attribute.MinimumLength.ToString(CultureInfo.InvariantCulture);
         }
 
-        public override IEnumerable<ModelClientValidationRule> GetClientValidationRules(
-            ClientModelValidationContext context)
+        /// <inheritdoc />
+        public override void AddValidation(ClientModelValidationContext context)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var errorMessage = GetErrorMessage(context);
-            var rule = new ModelClientValidationStringLengthRule(errorMessage,
-                                                                 Attribute.MinimumLength,
-                                                                 Attribute.MaximumLength);
-            return new[] { rule };
+            MergeAttribute(context.Attributes, "data-val", "true");
+            MergeAttribute(context.Attributes, "data-val-length", GetErrorMessage(context));
+            MergeAttribute(context.Attributes, "data-val-length-max", _max);
```

Comment:
Why are you changing the semantics here? Previous code checked equivalent of `Attribute.MaximumLength != int.MaxValue` before adding `data-val-length-max`.